### ISSUE TITLE
Thumb lookup on first edit when using Upload/edit file

### ISSUE THREAD

**Issue by jongfeli** [yaronkoren/TinyMCE#3]
When adding a Picture the thumb lookup is not working on **MW 1.30.0** (It only does this on the first edit).

On **MW 1.27.4** & **1.29.2** it is working just fine. I am not sure if it is possible to hide the URL and only show the picture name. It is a little confusing for a user to see a URL pop up when the picture is added. 

## On MW 1.27.4& 1.29.2
The thumbnail is loaded but it is a litle confusing that the URL is shown.
![mw-1 27 4_mw-1 29 2](https://user-images.githubusercontent.com/6680027/38858145-3afea538-422b-11e8-8da2-01e58122fb0d.png)

## On MW-1.30.0
Everything looks fine but the thumb is not loaded.
![mw-1 30 0](https://user-images.githubusercontent.com/6680027/38858363-e24da2f8-422b-11e8-8fe5-9bc0afcc179f.png)

**Comment by duncan-crane** [comment]
Hi Felipe

Thank you for your good work testing out the image upload/edit functionality.  It is very helpful as clearly you have a good understanding of how wikis handle images.  

This functionality was a bit of a labour of Hercules for me given all the options for image upload and display that mediawiki provides, some of which are a little quirky.  I am a novice wiki author so don't always appreciate or have use for many of these features, so its great to have someone testing that does.

I am also a 'weekend' coder so it may take me a little while to work through all of your comments so please do bear with me :-)

I have hopefully made the change you suggested in the second part of this report and have uploaded this to the master branch.  If you have a chance to test this please do let me know if this is now OK fro you?

For the first part of your report, I do not have MW 1.30 available on my system unfortunately.  I looked at the release notes for the intervening versions and couldn't see any reference to changes affecting the handling of Thumbnails so I wonder if you might provide some more details as to exactly what you have input and what the results were?

Many thanks

Duncan

**Comment by jongfeli** [comment]
Hello Duncan, I very much appreciate the time you and Yaron put in this extension or any extension for that matter. I am not in a position to demand anything I am just testing it, I actually like doing that, so there is no hurry :-)

I did try master but the URL still shows in the Source filename, there is no change in the behavior (MW 1.27 & 1.29).

In MW 1.27 & 1.29 when you select `Upload type:  File in this wiki` and then put in an existing file in` Source filename` the url immediately pops up. It does not do that at all in MW 1.30 I can put in any filename even one that does not exist. It seems that in MW 1.30.0 it does not do the lookup at that moment. In MW 1.27 & 1.29 you get the following pop up message when the file does not exits. You do **NOT** get this pop up message in MW 1.30:

![1](https://user-images.githubusercontent.com/6680027/38919388-ba2c4ffc-42f0-11e8-812c-f4d676f5a0f8.png)

**Comment by duncan-crane** [comment]
Hi Felipe

That is interesting that the fix didn’t work for you.

Is it possible to check in the mw_upload plugin.js to see if the following line is found at the end of function destChange?  It is on or around line 470 and shouldn’t be present in the latest version.  It was this line that replaces the source file value with the url of the thumbnail.

*	alternateSrcCtrl.value(destinationFileDetails);

The problem with MW 1.30 may be to do with the technique I’m using to make the API lookup synchronous on the wiki for the existing file.  I’m using the following code to do this, and possibly this isn’t supported in the MW 1.30 environment.

                // see if file already on wiki and return details if it is
               $.when(getFileDetailsFromWiki(destinationFile), $.ready).then( function(a1){
                               destinationFileDetails = a1;
               });

There is another approach of making the ajax call in getFileDetailsFromWiki synchronous which I can try.  I’ll maybe be able to do this later today as it should be a smallish fix.

Duncan

**Comment by duncan-crane** [comment]
Hi Felipe

I think I now understand better what was happening with the full url in the source field.  If you try to upload a local file whose name already exists on the wiki, you are given the option of displaying the existing wiki file.  If you accept this then the source field displays the full url to the thumbnail file, not the page reference for the file on the wiki?

My earlier correction was for the case where the user chose to use an existing file from the wiki in the first place, but not for the case I describe above.  I have now put in a fix for this case on the master branch which I hope now fully solves this part of the issue?

I've also put in a change to try the alternative way of making the API ajax call synchronous (as described in my earlier email).  I do not have MW 1.30, so I wondered if you might test this change?  It appears to work on  my MW 1.27 version OK.

Many thanks - Duncan

**Comment by jongfeli** [comment]
Hello Duncan, it now works on all MW versions. Thanks.

The only "problem" left is that under MW-1.30.0 you need to press the OK button twice to insert the picture on the page.

**Comment by duncan-crane** [comment]
I have made some significant changes to how uploaded files are displayed.  Rather than creating the HTML for displaying images in the extension, I now use the wiki parser to generate the html from the wiki code.  

For some reason I have had to recreate the mw css in MW_tinymce.css so there may still be some display issues if I've missed some definitions.  

Hopefully these changes haven't introduced any new bugs! They have simplified the code of the mw_wikicode and mw_upload plugins (although there is still some more tidying  up I need to do to complete this).  They should also have made the TinyMCE editor display look more like the rendered wiki page.

In the process I think I have also fixed the problem with having to click the OK button twice which in fact was happening for all MW versions, I think.
